rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,100 @@
+use crate::logger::{self, Level};
+use std::fs::{self, File, OpenOptions};
+use std::io::{Error, Read, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process;
+
+fn remove_lock(lock_path: &Path) {
+    fs::remove_file(lock_path).ok();
+    log!(
+        Level::Debug,
+        "FsLock: removed lockfile {}",
+        lock_path.display()
+    );
+}
+
+#[derive(Default)]
+pub struct FsLock {
+    lock_path: PathBuf,
+    lock_file: Option<File>,
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.lock_file.is_some() {
+            remove_lock(&self.lock_path);
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock(&mut self, new_lock_path: &Path, pid: &mut libc::pid_t) -> bool {
+        if self.lock_file.is_some() && self.lock_path == new_lock_path {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_path.display()
+        );
+
+        // first we open (and possibly create) the lock file
+        let mut options = OpenOptions::new();
+        options.create(true).read(true).write(true).mode(0o600);
+        let mut file = match options.open(&new_lock_path) {
+            Ok(file) => file,
+            Err(_) => return false,
+        };
+
+        // then we lock it (returns immediately if locking is not possible)
+        if unsafe { libc::lockf(file.as_raw_fd(), libc::F_TLOCK, 0) } == 0 {
+            log!(
+                Level::Debug,
+                "FsLock: locked `{}', writing PID...",
+                new_lock_path.display()
+            );
+            let pid = process::id().to_string();
+            let success = match unsafe { libc::ftruncate(file.as_raw_fd(), 0) } {
```

Comment:
I believe truncation can be done with safe Rust: `file.set_len(0)` ([docs on `set_len()`](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len)). Can you try that, please?

Reply:
Ah, I didn't know about this.